### PR TITLE
BONNIE-1072 VSAC Call Changes

### PR DESCRIPTION
**Pull request by alexanderelliott121**
Intent of PR is to bring the HDS vsac API into conformity with the vsac web api.

JIRA: https://jira.mitre.org/browse/BONNIE-1072
Tests: Automated

- [X] This pull request describes *why* these changes were made.
- [X] Tests are included and test edge cases
- [X] Link to JIRA test(s), if used: NA
- [X] JIRA test(s) have been added to sprint, if used NA
- [X] JIRA ticket for this PR: 
- [X] JIRA ticket links to this PR.
- [x] Automated regression test(s) pass (if applicable): 

Reviewer 1: @hossenlopp
- [x] The code is maintainable
- [x] The code reuses code or resources within the project ecosystem where applicable
- [x] The code achieves its purpose
- [x] The tests appropriately test edge cases
- [x] The tests pass: ONLY PASS LOCALLY. Travis errors are possibly a code issue from outside this PR. 

Reviewer 2: @c-monkey
- [x] The code is maintainable
- [x] The code reuses code or resources within the project ecosystem where applicable
- [x] The code achieves its purpose
- [x] The tests appropriately test edge cases
- [x] The tests pass

### CODE REVIEW

File: lib/health-data-standards/util/vs_api.rb [hunk 1]
```diff
@@ -69,6 +71,9 @@ def get_valueset(oid, options = {}, &block)
         profile = options.fetch(:profile, nil)
         effective_date = options.fetch(:effective_date, nil)
         program_name = options.fetch(:program, nil)
+        if profile.nil? && include_draft
```
Comment (hossenlopp) [inline lib/health-data-standards/util/vs_api.rb]: Perhaps this should also check if it is an empty string.

File: test/unit/util/vs_api_test.rb [hunk 2]
```diff
@@ -69,6 +69,14 @@ def test_api_v2_with_include_draft_default_profile
     end
   end
   
+  def test_api_v2_with_include_draft_no_profile
+    valueset_xml = %{<?xml version="1.0" encoding="UTF-8"?><RetrieveValueSetResponse xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" cacheExpirationHint="2012-09-20T00:00:00-04:00" xmlns:nlm="urn:ihe:iti:svs:2008" xmlns="urn:ihe:iti:svs:2008"><ValueSet id="2.16.840.1.113883.11.20.9.23" version="Draft"></ValueSet></RetrieveValueSetResponse>}
```
Comment (hossenlopp) [inline test/unit/util/vs_api_test.rb]: This line isn't needed for this test.

File: test/unit/util/vs_api_test.rb [hunk 3]
```diff
@@ -69,6 +69,14 @@ def test_api_v2_with_include_draft_default_profile
...
+    api = HealthDataStandards::Util::VSApiV2.new("https://localhost/token", "https://localhost/vsservice", "myusername", "mypassword")
+    assert_raises HealthDataStandards::Util::MalformedVSQueryError do
+      api.get_valueset("oid", include_draft: true)
+    end
+  end
+  
   def test_api_v2_with_include_draft_specified_profile
```
Comment (hossenlopp) [inline test/unit/util/vs_api_test.rb]: This existing test does execute the case where both `includeDraft` is `true` and `profile` is provided. It should, if possible, assert that the MalformedVSQueryError isn't thrown.

I noticed an existing issue with this test. All of the asserts for this test are in the block provided to get_valueset. If get_valueset returns early due to an introduced defect, then it won't yield anything and no asserts will occur. Therefore, this test may fail detect defects. The return result of the `get_valueset` function should be used to check the result of the test call.

**Review by hossenlopp — CHANGES_REQUESTED** [review]
I like the addition of the MalformedVSQueryError. It makes sure a hidden default doesn't accidentally get used. There's a few small checks and tests that could be improved.

**Comment by hossenlopp** [comment]
Further suggestion I forgot earlier: The `DEFAULT_PROFILE` should be removed since it's now unused.

File: lib/health-data-standards/util/vs_api.rb (hunk 1 above)
Comment (alexanderelliott121) [reply lib/health-data-standards/util/vs_api.rb]: I think that the options.fetch(:profile, nil) returns a nil if there is an empty string.
I'll do some testing.

**Comment by c-monkey** [comment]
I agree with all of @hossenlopp comments. Definitely put in some negative tests. Also the JIRA ticket says "It doesn't make sense for HDS to have a default profile...It's Bonnie's business to have a default profile." So might as well clean up the function description along with the DEFAULT_PROFILE.

Thanks!

**Review by c-monkey — CHANGES_REQUESTED** [review]
I agree with all of @hossenlopp comments. Definitely put in some negative tests. Also the JIRA ticket says "It doesn't make sense for HDS to have a default profile...It's Bonnie's business to have a default profile." So might as well clean up the function description along with the DEFAULT_PROFILE.

Thanks!

File: test/unit/util/vs_api_test.rb (hunk 3 above)
Comment (alexanderelliott121) [reply test/unit/util/vs_api_test.rb]: I'm going to create a separate JIRA for this, since it would change the scope of the changes, but I agree that is should be addressed.

File: lib/health-data-standards/util/vs_api.rb [hunk 4]
```diff
@@ -59,8 +59,6 @@ class VSApiV2 < VSApi
       # The VSAC V2 API needs a profile to be specified when using includeDraft. Future work on this
```
Comment (hossenlopp) [inline lib/health-data-standards/util/vs_api.rb:58]: At least the first line of the comment should be removed since you removed the `DEFAULT_PROFILE`.

**Comment by alexanderelliott121** [comment]
@c-monkey I'm not sure I follow what you mean by negative test cases.
Currently I test to make sure that the correct exception is thrown if the wrong arguments are provided.  Is there another negative case I'm missing?

**Review by hossenlopp — APPROVED** [review]

File: test/unit/util/vs_api_test.rb (hunk 3 above)
Comment (c-monkey) [reply test/unit/util/vs_api_test.rb]: Please add the link to the separate JIRA.

**Review by c-monkey — CHANGES_REQUESTED** [review]
We decided during the sprint review that if it is expedient, you should just finish up the testing before this PR is accepted. I propose only a few changes:

1) Following up on the request from @hossenlopp I did a little researching, and there is no such thing as asserting that an error wasn't thrown. True, we could create a try/catch block, BUT, if the error were thrown in another test, it would trigger an error, so it's not necessary to test. If desired, we could add a comment to the affect of: `# does not raise HealthDataStandards::Util::MalformedVSQueryError` or somesuch into test test_api_v2_with_include_draft_default_profile.

2) These tests miss a little of the logic, and could be revised to capture all logic permutations, but `get_valueset` is a very simple function. The tests are using the implicit matching of the parameters to the the call to `get_valueset` and the expected parameters in the `.with` of the `stub_request`. Since I was new to the WebMock library, didn't know there was no way to inspect parameters that were sent to the mocked URLs, so this use of WebMock is pretty clever.

3) For completeness, one could test that the stub *isn't* called, i.e.:
```
  def test_api_v2_with_include_draft_no_profile
    the_stub = stub_request(:get,'https://localhost/vsservice')
    api = HealthDataStandards::Util::VSApiV2.new("https://localhost/token", "https://localhost/vsservice", "myusername", "mypassword")
    assert_raises HealthDataStandards::Util::MalformedVSQueryError do
      api.get_valueset("oid", include_draft: true)
    end
    assert_not_requested the_stub
  end
```
Do you have an objection to this? I think it adds clarity.

Thanks,
--Chris

p.s. all tests in HDS and in Bonnie pass.

**Comment by edeyoung** [comment]
What's the status on the regression tests?

**Comment by c-monkey** [comment]
@alexanderelliott121 The travis tests are failing identically to previous PRs. In a previous PR @edeyoung pointed out that it was failing on some people's machines (but not yours or hers, and now this one also doesn't fail on my machine).

Therefore I'm okay with integrating this change.

**Review by c-monkey — APPROVED** [review]

**Comment by c-monkey** [comment]
Travis failure ticket is : https://jira.mitre.org/browse/BONNIE-1136

File: test/unit/util/vs_api_test.rb [hunk 5]
```diff
@@ -63,6 +63,7 @@ def test_api_v2_with_include_draft_default_profile
     valueset_xml = %{<?xml version="1.0" encoding="UTF-8"?><RetrieveValueSetResponse xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" cacheExpirationHint="2012-09-20T00:00:00-04:00" xmlns:nlm="urn:ihe:iti:svs:2008" xmlns="urn:ihe:iti:svs:2008"><ValueSet id="2.16.840.1.113883.11.20.9.23" version="Draft"></ValueSet></RetrieveValueSetResponse>}
     stub_request(:get,'https://localhost/vsservice').with(:query =>{:id=>"oid", :ticket=>"ticket", :includeDraft=>"yes", :profile=>"Most Recent CS Versions"}).to_return(:body=>valueset_xml)
     api = HealthDataStandards::Util::VSApiV2.new("https://localhost/token", "https://localhost/vsservice", "myusername", "mypassword")
+    # does not raise HealthDataStandards::Util::MalformedVSQueryError
     api.get_valueset("oid", include_draft: true, :profile=>"Most Recent CS Versions") do |oid,vs|
```
Comment (hossenlopp) [inline test/unit/util/vs_api_test.rb]: The result of calling get_valuset should be checked instead of asserting stuff in the block. If get_valueset doesn't actually attempt to get the valueset and just returns nothing, this test won't catch that since the block won't get called at all.

**Review by hossenlopp — CHANGES_REQUESTED** [review]

File: test/unit/util/vs_api_test.rb (hunk 5 above)
Comment (c-monkey) [reply test/unit/util/vs_api_test.rb]: This is a problem in all the tests, then. I know you (@hossenlopp) caught this before but when debugging how the strange stub_request code works, I forgot about this very base case. Since the first parameter `oid` is not changed, we could make a simple test for testing the yield and then in all the rest of these tests just use the normal return value. Removing the, now basically worthless `assert_equal "oid", oid` assertions would be nice for cleaning up the code, too.

So, just change all the calls to `get_valueset` to `vs = ...` and remove the block. Then the one assertion in the block out-dented.

Comment (alexanderelliott121) [reply test/unit/util/vs_api_test.rb]: Basically, I think making that change should constitute its own ticket, given its scope.

Comment (c-monkey) [reply test/unit/util/vs_api_test.rb]: @hossenlopp 's suggested refactor is trivial, changing 4 tests, each only 2 lines, and identical changes. I disagree that this should be taken out of this PR. The new Bonnie Way is to fix broken tests before we commit.

File: test/unit/util/vs_api_test.rb (hunk 2 above)
Comment (alexanderelliott121) [reply test/unit/util/vs_api_test.rb]: I needed to put it back in for the assert_not_requested.

**Comment by alexanderelliott121** [comment]
Tests have been updated.

**Review by c-monkey — APPROVED** [review]
Tested bonnie as well

**Review by hossenlopp — APPROVED** [review]
